### PR TITLE
Add xpu legacy profiler for Intel(R) XPU devices

### PR DESCRIPTION
**Pull request by xunsongh**
# Motivation
This patch aimed to add an xpu path of legacy profiler for Intel(R) XPU devices. With this PR,  users will be able to profile GPU kernel time consuming and memory allocations out-of-queue while using Intel(R) XPU backend.

# Solution
We add use_xpu for entering xpu path of legacy profiler and we imitate CUDA to designed how xpu_time_total and self_xpu_time_total should be calculated. Different from CUDA and CPU who use 'push' range and 'pop' range to get the interval of an operator, we designed to use 'mark' to mark the XPU kernels to the profiler. One XPU kernel will calculate its elapsed time by the SYCL event related to the operator itself.
Also, we redefined the struct of KernelEventBase, which can be used as a template for all backends to design their own profiler in a unified way, not only for CUDA or XPU, but also for other backends which will raise independent kernels while running and profiling.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/autograd/profiler_util.py [hunk 1]
```diff
@@ -487,15 +506,47 @@ def cpu_time_total(self):
...
 
+    @property
+    def xpu_time_total(self):
+        if self.is_async:
+            return 0
+        if self.device_type == DeviceType.CPU:
+            if not self.is_legacy:
+                # account for the kernels in the children ops
+                if len(self.kernels) > 0:
+                    assert self.kernels[0].device.type == DeviceType.CUDA
```
Comment (chengjunlu) [inline torch/autograd/profiler_util.py:517]: DeviceType.XPU?

File: torch/csrc/autograd/profiler_legacy.h [hunk 2]
```diff
@@ -209,13 +274,24 @@ struct TORCH_API LegacyEvent {
...
 
+  bool hasKernel() const {
+    return kernel_event_ != nullptr || (isRemote() && device_.index() != -1);
+  }
+
   double cudaElapsedUs(const LegacyEvent& e) const;
 
   bool hasCuda() const {
-    return cuda_event != nullptr || (isRemote() && device_ != -1);
+    return hasKernel() && device_.type() == DeviceType::CUDA;
```
Comment (chengjunlu) [inline torch/csrc/autograd/profiler_legacy.h:284]: May the device type be CPU and the kernel is for the CUDA?

**Comment by gujinghui** [comment]
@ezyang  Could you help review this PR? Thanks.

**Comment by gujinghui** [comment]
Hi @ezyang

Could you help review this PR?

Thanks.

**Comment by ezyang** [comment]
I suppose it has to be asked... why are we adding new functionality to the legacy profiler?

**Comment by gujinghui** [comment]
> I suppose it has to be asked... why are we adding new functionality to the legacy profiler?

Hi @ezyang ,

We are going to support both legacy profiler and kineto.
This PR is only for legacy profiler. And we will plan to support kineto in next step.

The motivation of supporting legacy is, we still have many users who are using legacy profiler, both internal and external.
Therefore, we need to support it, as well.

Thanks,

**Comment by gdankel** [comment]
@gujinghui as a general rule we won't accept patches to the legacy profiler. Can you elaborate on why people are still using the legacy profiler? I believe we have feature parity with Kineto. If there are any other reasons please let us know and we will address them. We do not want to continue development on the legacy profiler.

**Comment by ezyang** [comment]
(unsubscribing myself; @gdankel can you drive this to conclusion?)

**Pull request closed without merging**